File: task-frontend/src/mvc/model/itineraryModel.js
/**
 * itineraryModel.js — Data model for itinerary state management
 */

/**
 * Creates a default empty itinerary state object.
 * @returns {Object}
 */
export function createEmptyItinerary() {
  return {
    days: [],
  };
}

/**
 * Validates that the itinerary response has the expected shape.
 * @param {Object} data
 * @returns {boolean}
 */
export function isValidItinerary(data) {
  return (
    data &&
    Array.isArray(data.days) &&
    data.days.length > 0 &&
    data.days.every(
      (day) =>
        typeof day.day === "number" &&
        typeof day.title === "string" &&
        typeof day.area_focus === "string" &&
        typeof day.summary === "string" &&
        Array.isArray(day.stops) &&
        day.stops.length > 0 &&
        day.stops.every(
          (s) =>
            typeof s.time === "string" &&
            typeof s.place_name === "string" &&
            typeof s.details === "string"
        )
    )
  );
}

File: task-backend/app/models/itinerary_model.py
from pydantic import BaseModel
from typing import List, Optional

class ItineraryRequest(BaseModel):
    destination: str
    style: str


# EXPLAIN: We return a structured itinerary (instead of freeform strings) so the
# UI can render a timeline, show images per stop, and stay consistent across
# destinations/styles.
# EXPLAIN: Use plain strings (not strict Literals) so the API doesn't crash if
# the model returns slight variations like "Late afternoon" vs "Late Afternoon".
# The prompt still strongly constrains the allowed values; this just improves
# runtime reliability (no more 500s from response validation).
StopTime = str
StopCategory = str


class StopImage(BaseModel):
    image_url: str
    source: str
    page_url: Optional[str] = None


class DayStop(BaseModel):
    time: StopTime
    place_name: str
    place_query: str
    category: StopCategory
    duration_minutes: int
    details: str
    pro_tip: str
    image: Optional[StopImage] = None


class DayPlan(BaseModel):
    day: int
    title: str
    area_focus: str
    summary: str
    stops: List[DayStop]
    must_try_foods: List[str]
    logistics: List[str]

class ItineraryResponse(BaseModel):
    days: List[DayPlan]

class ErrorResponse(BaseModel):
    error: str
